### ISSUE TITLE
Route Resolver Issue

### ISSUE THREAD

**Issue by juliocgm83** [MarkPieszak/aspnetcore-angular2-universal#455]
I am using a resolver in a route, everything work correctly when I go to the route using a navigate function but once I am in the page of that route and I reload the page by pressing F5 I got this error:

**NodeInvocationException: Prerendering timed out after 30000ms because the boot function in 'C:\Test/ClientApp/dist/main-server' returned a promise that did not resolve or reject. Make sure that your boot function always resolves or rejects its promise. You can change the timeout value using the 'asp-prerender-timeout' tag helper.**

Any Idea ?

**Comment by isaac2004** [comment]
@juliocgm83 are you running this in IIS or Kestrel? You should get a prerending error in the Output tab of VS or the Kestrel window

**Comment by juliocgm83** [comment]
@isaac2004 Yes I am using .net core and kestrel, the error that I am getting in the output is the same that I posted before, I am including now part of the call stack referring to the error.

PrekClub.Server.Helpers.HttpRequestExtensions+<BuildPrerender>d__1.MoveNext() in HttpRequestExtensions.cs
-
			//Prerender now needs CancellationToken
			System.Threading.CancellationTokenSource cancelSource = new System.Threading.CancellationTokenSource();
			System.Threading.CancellationToken cancelToken = cancelSource.Token;
			// Prerender / Serialize application (with Universal)
			return await Prerenderer.RenderToString(
				"/",
				nodeServices,
				cancelToken,
				new JavaScriptModuleExport(applicationBasePath + "/ClientApp/dist/main-server"),
				unencodedAbsoluteUrl,
				unencodedPathAndQuery,
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
System.Runtime.CompilerServices.TaskAwaiter.GetResult()
PrekClub.Server.Controllers.HomeController+<Index>d__2.MoveNext() in HomeController.cs
-
        {
	        var dataToTransfer = new DefaultViewModel()
	        {
		        GlobalSetting = _globalSetting
	        };
			var prerenderResult = await Request.BuildPrerender(dataToTransfer);
			ViewData["SpaHtml"] = prerenderResult.Html; // our <app> from Angular
            ViewData["Title"] = prerenderResult.Globals["title"]; // set our <title> from Angular
            ViewData["Styles"] = prerenderResult.Globals["styles"]; // put styles in the correct place
            ViewData["Scripts"] = prerenderResult.Globals["scripts"]; // scripts (that were in our header)
            ViewData["Meta"] = prerenderResult.Globals["meta"]; // set our <meta> SEO tags
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

**Comment by juliocgm83** [comment]
@isaac2004 I made a litle change to the resolver to return null when it is server platform and now is working, but I don't know if that is a good solution 

original resolver code:
resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): IPlace | Observable<IPlace> | Promise<IPlace> {
	         const placeId = route.paramMap.get('placeId');
	        return this.gmapService.getPlace(placeId).take(1).map((place: IPlace) => {
		if (!place) {
			this.router.navigate(['']);
			return null;
		} else {
			return place;
		}
	});
}

resolver code now:
resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): IPlace | Observable<IPlace> | Promise<IPlace> {
	if (isPlatformBrowser(this.platformId)) {
		const placeId = route.paramMap.get('placeId');
		return this.gmapService.getPlace(placeId).take(1).map((place: IPlace) => {
			if (!place) {
				this.router.navigate(['']);
				return null;
			} else {
				return place;
			}
		});
	} else {
		return null;
	}
	
}

**Comment by isaac2004** [comment]
Here is a resolver I am using in another repo and it works completely fine. You should be able to see what the JS error that is causing the issue if you read the output carefully, it will be before the timeout error

````
import { Injectable } from '@angular/core';
import {
    Router, Resolve, RouterStateSnapshot,
    ActivatedRouteSnapshot
} from '@angular/router';
import { Observable } from 'rxjs/Observable';
import { AboutService } from './about.service';
import { LoadingService } from './loading.service';


@Injectable()
export class AboutResolver implements Resolve<any> {
    constructor(private as: AboutService, private router: Router, public loadingService: LoadingService) { }

    resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<any> {
        this.loadingService.loaded = false;
        return this.as.get().map(about => {
            if (about) {
                this.loadingService.loaded = true;
                return about;
            } else { // id not found
                this.router.navigate(['']);
                return undefined;
            }
        });
    }
}

````

**Comment by juliocgm83** [comment]
@isaac2004 
I found this issue in the output before the time out

Microsoft.AspNetCore.NodeServices[0]
      ERROR { Error: Uncaught (in promise): TypeError: __WEBPACK_IMPORTED_MODULE_2_rxjs_Observable__.Observable.throw is not a function
      TypeError: __WEBPACK_IMPORTED_MODULE_2_rxjs_Observable__.Observable.throw is not a function
          at CatchSubscriber.selector (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\main-server.js:45844:85)
          at CatchSubscriber.module.exports.CatchSubscriber.error (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:84945:31)
          at DoSubscriber.module.exports.DoSubscriber._error (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:92259:30)
          at DoSubscriber.module.exports.Subscriber.error (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:660:18)
          at MapSubscriber.module.exports.Subscriber._error (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:686:26)
          at MapSubscriber.module.exports.Subscriber.error (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:660:18)
          at ZoneTask.onComplete [as callback] (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\main-server.js:42752:30)
          at ZoneDelegate.module.exports.ZoneDelegate.invokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66170:31)
          at Object.onInvokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\main-server.js:4161:33)
          at ZoneDelegate.module.exports.ZoneDelegate.invokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66169:36)
          at resolvePromise (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66569:31)
          at resolvePromise (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66540:17)
          at C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66618:17
          at ZoneDelegate.module.exports.ZoneDelegate.invokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66170:31)
          at Object.onInvokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\main-server.js:4161:33)
          at ZoneDelegate.module.exports.ZoneDelegate.invokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66169:36)
          at Zone.module.exports.Zone.runTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:65937:47)
          at drainMicroTaskQueue (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66347:35)
          at module.exports.ZoneTask.invokeTask (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66248:21)
          at ZoneTask.invoke (C:\Work\www\PrekClub\DIrectoryV2\ClientApp\dist\vendor.js:66233:48)

**Comment by juliocgm83** [comment]
@isaac2004 
will it be missing the import of the throw import 'rxjs/add/observable/throw' in main-server.js?

**Comment by isaac2004** [comment]
are you catching or throwing anywhere in your app? If so, you may need to explicitly declare that module

**Comment by juliocgm83** [comment]
@isaac2004 
I made some tests and this is what I got:
The first issue was because I have an implementation of the http and i was catching any error and after that throw again an exception like this Observable.throw...look like that the server side does not know the throw function of the onservable, I remove that and I start getting this error: 
**ERROR { Error: Uncaught (in promise): Response with status: 0  for URL: null**
After that to be sure was not problem of the my http service I change the code to use directly the http service provide by angular but I am still getting the same error. I am thinking it could be an issue with cross domain requests or something because my api is in other domain, but I am not sure because from the client side it is working perfectly and I allow in the api request from all origins.
Sorry for bother you so much, I am doing my first steps in this technology

**Comment by isaac2004** [comment]
Are you hitting the web service at all? Are you getting a response?

**Comment by juliocgm83** [comment]
look like it is something with the server request to my api because I put a call to the a service in the home component like this:
import { Component, OnInit } from '@angular/core';
import { GmapService } from "../../shared/services/test.service";


@Component({
    selector: 'app-home',
	templateUrl: './home.component.html',
    styleUrls: ['./home.component.scss']
})
export class HomeComponent implements OnInit {

    title: string = 'Angular 4.0 Universal & ASP.NET Core 2.0 advanced starter-kit';

    // Use "constructor"s only for dependency injection
    constructor(private service: TestService
     
    ) { }

    // Here you want to handle anything with @Input()'s @Output()'s
    // Data retrieval / etc - this is when the Component is "ready" and wired up
	ngOnInit() {
		this.service.getData('ChIJwTkBpwrB2YgROJuU0W6dDH8').subscribe((response) => {
			console.log('response => ', response);
		});
	}

}

and from the client it is working perfectly but tin the output of the server I got an error response like this:
 Microsoft.AspNetCore.NodeServices[0]
      ERROR Response {
        _body:
         ProgressEvent {
           type: 'error',
           target:
            XMLHttpRequest {
              onloadstart: null,
              onprogress: null,
              onabort: null,
              onerror: null,
              onload: null,
              ontimeout: null,
              onloadend: null,
              _listeners: [Object],
              onreadystatechange: null,
              _anonymous: undefined,
              readyState: 4,
              response: null,
              responseText: '',
              responseType: '',
              responseURL: '',
              status: 0,
              statusText: '',
              timeout: 0,
              upload: [Object],
              _method: 'GET',
              _url: [Object],
              _sync: false,
              _headers: [Object],
              _loweredHeaders: [Object],
              _mimeOverride: null,
              _request: null,
              _response: null,
              _responseParts: null,
              _responseHeaders: null,
              _aborting: null,
              _error: null,
              _loadedBytes: 0,
              _totalBytes: 0,
              _lengthComputable: false },
           currentTarget:
            XMLHttpRequest {
              onloadstart: null,
              onprogress: null,
              onabort: null,
              onerror: null,
              onload: null,
              ontimeout: null,
              onloadend: null,
              _listeners: [Object],
              onreadystatechange: null,
              _anonymous: undefined,
              readyState: 4,
              response: null,
              responseText: '',
              responseType: '',
              responseURL: '',
              status: 0,
              statusText: '',
              timeout: 0,
              upload: [Object],
              _method: 'GET',
              _url: [Object],
              _sync: false,
              _headers: [Object],
              _loweredHeaders: [Object],
              _mimeOverride: null,
              _request: null,
              _response: null,
              _responseParts: null,
              _responseHeaders: null,
              _aborting: null,
              _error: null,
              _loadedBytes: 0,
              _totalBytes: 0,
              _lengthComputable: false },
           lengthComputable: false,
           loaded: 0,
           total: 0 },
        status: 0,
        ok: false,
        statusText: '',
        headers: Headers { _headers: Map {}, _normalizedNames: Map {} },
        type: 3,
        url: null }

**Comment by juliocgm83** [comment]
@isaac2004 Finally I figured it out
The issue is related to the ssl certificate of my local pc, I was working with https and look like there something wrong with the development certificate that I have installed. I just jump to http and it work fine. Tomorrow I will test using the test server which have a valid ssl certificate to be sure the issue is only in my local pc.
Thanks so much for your help
